synthesis/power_performance_area.proto

```diff
@@ -0,0 +1,165 @@
+
+  // A fraction of die_area_um2 used by cells. Range of [0..1]
+  optional float cell_utilization_fraction = 5;
+
+  // Leaving out proto tags 6..9 for later use.
+
+  // The total number of standard cells used.
+  // Invariance:
+  // total_cells = buffers + timing_buffers + flops + combinational;
+  optional int32 num_total_cells = 10;
+
+  // The number of combinational gate cells.
+  optional int32 num_combinational = 11;
+
+  // The number of non-timing buffers.
+  optional int32 num_buffers = 12;
+
+  // The number of timing buffers.
+  optional int32 num_timing_buffers = 13;
+
+  // The number of flops.
+  optional int32 num_flops = 14;
+
+  // Leaving out proto tags 15..19 for later use.
+
+  // Cell type mapping to fraction used in the area, e.g.
+  // {{ "svt: 0.7 }, { "lvt": 0.3}}
+  map<string, float> celltype_fractions = 20;
+}
+
+message Performance {
+  // Desired clock frequency this PPA metric was generated for.
+  // It influences the choice of area-impacting optimization, so this
+  // helps to understand area size in that context.
+  optional int32 clock_period_ps = 1;
+
+  // Longest path in picoseconds between any two flop-boundaries.
+  optional int32 critical_path_ps = 2;
+
+  // == Slack ==
+  // Abbreviations
+  // wns = worst negative slack (worst negative slack observed in all paths)
+  //       (encoded as sint32 as these are always negative and zigzag encoding
+  //        is preferable)
+  // tns = total negative slack (sum of slacks over all paths that have
+  //      negative slack). Can get big, so sint64
+  //
+  // Slack is the amount we have leftover in some clock period,
+  // so if positive, we're good. Negative slack is the one interesting
+  // to evaluate timing. In fact, tools rarely report a positive slack
+  // just simply zero as 'mission accomplished'.
+  // So the following numbers are negative if there is anything to
+  // worry about.
+  // Approximate invariant:
+  //   clock_period_ps - critical_path_ps = setup_wns_ps
+
+  // If negative, not enough time to have data ready for DFF to take data, i.e.
+  // the combinational path that exceeds that time the most.
+  optional sint32 setup_wns_ps = 3;
+  optional sint64 setup_tns_ps = 4;
+
+  // If negative, new data arrives too early for the DFF to ingest
+  // the previous one reliably.
+  optional sint32 hold_wns_ps = 5;
+  optional sint64 hold_tns_ps = 6;
+
+  // How far off the clock is between two different parts of the chip.
+  optional sint32 clock_skew_ps = 7;
+
+  // Leaving out proto tags 8..9 for later use.
+
+  optional int32 num_setup_violations = 10;
+  optional int32 num_hold_violations = 11;
+  optional int32 num_slew_violations = 12;
+  optional int32 num_fanout_violations = 13;
+  optional int32 num_capacitance_violations = 14;
+
+  // The number of cells in the longest topological path.
+  optional int32 longest_topological_path = 15;
+
+  // Number of cells in the critical path
+  optional int32 critical_path_cells = 16;
```

Comment:
How is this different than above? Does longest path mean in-to-out? If so, I would specify that int he comment, as a I was assuming "longest_topological_path" was a constrained path.

Reply:
Clarified in comment that the criticial path one is based on timing.